### PR TITLE
Fix a bug in autodiff caused by the implicit capture of lower_random

### PR DESCRIPTION
**Pull request by BachiLi**
This PR fixes a bug in autodiff when using with `random_float` or `random_int`. During `lower_random`, the implicit capture creates references to reduction variables without defining their `reduction_domain`, causing issues in autodiff. I fixed this by reintroducing the reduction domain when required. I also removed a few warnings when using random_float caused by propagating the derivative through some bit manipulation intrinsic functions.

### CODE REVIEW

**Comment by abadams** [comment]
Wouldn't it be better to not drop the domain reference in the first place? I.e. in Function.cpp make the free_vars vector a `vector<VarOrRVar>` instead of a `vector<string>`? It looks like it would actually slightly simplify LowerRandom, which takes a vector of string and reconstructs a vector of Var

**Comment by BachiLi** [comment]
> Wouldn't it be better to not drop the domain reference in the first place? I.e. in Function.cpp make the free_vars vector a `vector<VarOrRVar>` instead of a `vector<string>`? It looks like it would actually slightly simplify LowerRandom, which takes a vector of string and reconstructs a vector of Var

Good point. Made the suggested changes.

File: src/Function.cpp [hunk 1]
```diff
@@ -427,8 +428,13 @@ void Function::define(const vector<string> &args, vector<Expr> values) {
 
     // Tag calls to random() with the free vars
     int tag = rand_counter++;
+    vector<VarOrRVar> free_vars;
+    free_vars.reserve(args.size());
```
Comment (abadams) [inline src/Function.cpp]: Why not:
```
vector<VarOrRVar> free_vars(args.size());
```

File: src/Function.cpp [hunk 2]
```diff
@@ -595,16 +601,20 @@ void Function::define_update(const vector<Expr> &_args, vector<Expr> values) {
     }
 
     // Tag calls to random() with the free vars
-    vector<string> free_vars;
+    vector<VarOrRVar> free_vars;
+    int num_free_vars = (int)pure_args.size();
+    if (check.reduction_domain.defined()) {
+        num_free_vars += (int)check.reduction_domain.domain().size();
+    }
+    free_vars.reserve(num_free_vars);
```
Comment (abadams) [inline src/Function.cpp]: similarly:
```
vector<VarOrRVar> free_vars(num_free_vars);
```

File: src/Function.cpp [hunk 3]
```diff
@@ -427,8 +428,13 @@ void Function::define(const vector<string> &args, vector<Expr> values) {
 
     // Tag calls to random() with the free vars
     int tag = rand_counter++;
+    vector<VarOrRVar> free_vars;
+    free_vars.reserve(args.size());
+    for (size_t i = 0; i < args.size(); i++) {
+        free_vars.emplace_back(Var(args[i]));
```
Comment (abadams) [inline src/Function.cpp]: then free_vars[i] = Var(args[i])

File: src/Function.cpp (hunk 2 above)
Comment (abadams) [reply src/Function.cpp]: Oh, no default constructor on VarOrRVar?

File: src/Function.cpp [hunk 4]
```diff
@@ -427,8 +428,13 @@ void Function::define(const vector<string> &args, vector<Expr> values) {
 
     // Tag calls to random() with the free vars
     int tag = rand_counter++;
+    vector<VarOrRVar> free_vars;
+    free_vars.reserve(args.size());
+    for (size_t i = 0; i < args.size(); i++) {
```
Comment (abadams) [inline src/Function.cpp]: Prefer range-based for loops when the index isn't used:
```
for (const auto &arg : args) {
  free_vars.emplace_back(Var(arg));
}
```

Comment (BachiLi) [reply src/Function.cpp]: I propagated the changes to a few loops in proximity. Let me know if I changed too much.

**Comment by abadams** [comment]
Thanks for cleaning up more of the for loops

**Comment by BachiLi** [comment]
@steven-johnson FYI the windows builtbots are having issues.

**Pull request merged**